src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,172 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> tail;
+    private Node<T> head;
+    private int size;
+
+    public MyLinkedList() {
+        head = null;
+        tail = null;
+    }
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(tail, null, value);
+        if (head == null) {
+            head = newNode;
+        }
+        if (tail != null) {
+            tail.next = newNode;
+        }
+        tail = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size()) {
+            add(value);
+            return;
+        }
+        testIndex(index);
+        if (index == 0) {
+            Node<T> newNode = new Node<>(null, head, value);
+            head.prev = newNode;
+            head = newNode;
+        } else {
+            Node<T> currentNode = getNode(index);
+            Node<T> newNode = new Node<>(currentNode.prev, currentNode, value);
+            currentNode.prev.next = newNode;
+            currentNode.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        testIndex(index);
+        Node<T> pointer = getNode(index);
+        if (pointer == null) {
+            return null;
+        } else {
+            return pointer.getValue();
+        }
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        testIndex(index);
+        int count = 0;
+        Node<T> pointer = head;
+        T oldValue = null;
+        while (pointer != null) {
+            if (index == count) {
+                oldValue = pointer.getValue();
+                pointer.setValue(value);
+                break;
+            } else {
+                pointer = pointer.next;
+                count++;
+            }
+        }
```

Comment:
you have the same logic inside getNode(int index) method
```suggestion
        Node<T> pointer = getNode(index);
        T oldValue = null;
        if (pointer != null) {
           oldValue = pointer.value;
           pointer.value = value;
        }
```